dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java

```diff
@@ -9,63 +9,141 @@
- * 
- * @author Andrea Bollini (andrea.bollini at 4science.it)
- *
  */
 @RestController
 @RequestMapping("/api/"+BitstreamRest.CATEGORY +"/"+ BitstreamRest.PLURAL_NAME + "/{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/content")
-public class BitstreamContentRestController {
+public class BitstreamContentRestController implements InitializingBean {
+
+	private static final Logger log = Logger.getLogger(BitstreamContentRestController.class);
+
+	@Autowired
+	private BitstreamService bitstreamService;
+
+	@Autowired
+	private EventService eventService;
+
 	@Autowired
-	private BitstreamRestRepository bitstreamRestRepository; 
-	
-	@RequestMapping(method = RequestMethod.GET)
+	private ConfigurationService configurationService;
+
+	@Autowired
+    private AuthorizeService authorizeService;
+
+	private int bufferSize;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        bufferSize = configurationService.getIntProperty("bitstream-download.buffer.size", -1);
+    }
+
+	@RequestMapping(method = {RequestMethod.GET, RequestMethod.HEAD})
 	public void retrieve(@PathVariable UUID uuid, HttpServletResponse response,
-            HttpServletRequest request) throws IOException {
-		BitstreamRest bit = bitstreamRestRepository.findOne(uuid);
-		if (bit == null) {
-			response.sendError(HttpServletResponse.SC_NOT_FOUND);
-			return;
+											 HttpServletRequest request) throws IOException, SQLException {
+
+		Context context = ContextUtil.obtainContext(request);
+
+        Bitstream bit = getBitstreamIfAuthorized(context, uuid, response);
+        if (bit == null) {
+            //The bitstream was not found or we're not authorized to read it.
+            return;
+        }
+
+        Long lastModified = bitstreamService.getLastModified(bit);
+        String mimetype = bit.getFormat(context).getMIMEType();
+
+		// Pipe the bits
+		try(InputStream is = getInputStream(context, bit)) {
+
+			MultipartFileSender sender = MultipartFileSender
+					.fromInputStream(is)
+					.withBufferSize(bufferSize)
+					.withFileName(bit.getName())
+					.withLength(bit.getSize())
+					.withChecksum(bit.getChecksum())
+					.withMimetype(mimetype)
+					.withLastModified(lastModified)
+					.with(request)
+					.with(response);
+
+			if (sender.isValid()) {
+				sender.serveResource();
+			}
+
+			if (sender.isNoRangeRequest() && isNotAnErrorResponse(response)) {
+				//We only log a download request when serving a request without Range header. This is because
+				//a browser always sends a regular request first to check for Range support.
+				eventService.fireEvent(
+						new UsageEvent(
+								UsageEvent.Action.VIEW,
+								request,
+								context,
+								bit));
+			}
+
+		} catch(IOException ex) {
+			log.debug("Client aborted the request before the download was completed. Client is probably switching to a Range request.", ex);
```

Comment:
are you sure about that?  an IOException can occur if we have issue to read from the underline storage. If the client abort the request the exception should occur on the Servlet container stack as writing to the response is in some way "buffered" by tomcat, etc. (I admit I haven't check)

Reply:
We'll make the check more specific on the "Client aborted" use case.